homeassistant/components/binary_sensor/webdav.py

```diff
@@ -0,0 +1,126 @@
+    vol.Optional(CONF_TOKEN): cv.string,
+    vol.Optional(CONF_CERTIFICATE_PATH): cv.string,
+    vol.Optional(CONF_KEY_PATH): cv.string,
+    vol.Optional(CONF_PATH, default="/"): cv.string,
+})
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up a single WebDAV share."""
+    from webdav3.client import Client
+    from webdav3.exceptions import WebDavException
+
+    name = config.get(CONF_NAME)
+    host = config.get(CONF_HOST)
+    path = config.get(CONF_PATH)
+    username = config.get(CONF_USERNAME)
+    password = config.get(CONF_PASSWORD)
+    token = config.get(CONF_TOKEN)
+    cert_path = config.get(CONF_CERTIFICATE_PATH)
+    key_path = config.get(CONF_KEY_PATH)
+    client = Client({
+        "webdav_hostname": host,
+        "webdav_login": username,
+        "webdav_password": password,
+        "webdav_token": token,
+        "webdav_cert_path": cert_path,
+        "webdav_key_path": key_path,
+        "webdav_root": path,
+    })
+
+    try:
+        client.list('/')  # This will throw if we can't access the share
+        add_entities([WebDAVSensor(name, client)], update_before_add=True)
+    except WebDavException as exception:
+        _LOGGER.error(
+            "Failed to connect to %s: %s",
+            client.get_url(""),
+            exception)
+
+
+class WebDAVSensor(BinarySensorDevice):
+    """A sensor that monitors a WebDAV share."""
+
+    def __init__(self, name, client):
+        """Create the sensor."""
+        self._name = name
+        self._client = client
+
+        self._files = []
+        self._available = False
+
+    def update(self):
+        """Get the current contents of the share."""
+        from webdav3.exceptions import WebDavException
+        try:
+            directory = '/'
+            self._files = [
+                self._client.get_url(directory + filename)
+                for filename in self._client.list(directory)
+                if not filename.endswith('/')]
+            self._available = True
+        except WebDavException as exception:
+            self._files = []
+            self._available = False
+            _LOGGER.debug(
+                "Could not open share %s. Message: %s",
+                self.unique_id,
+                exception)
+
+    @property
+    def name(self):
+        """Return the name of this sensor."""
+        return self._name
+
+    @property
+    def unique_id(self):
+        """Return the URL for the share."""
+        return self._client.get_url("")
+
+    @property
+    def is_on(self):
+        """Return true if the share is currently available."""
+        return self._available
+
+    @property
+    def device_class(self):
+        """Return the device class of this sensor."""
+        return DEVICE_CLASS
+
+    @property
+    def device_state_attributes(self):
+        """Return the list of files on the share."""
+        return {
+            "files": self._files,
```

Comment:
I'm concerned this list can become very big and blow up the UI... can we have a limit on the number of files here at least ? say 10 ?

Reply:
OK, you described a use case that is by no means related to this PR.

This binary_sensor is about connectivity (see your DEVICE_CLASS). It should be just about availability of the webdav server an nothing more.

What you wan't is an entity that feeds you files into your photo frame, right ? I think you could model your use case through a **camera**, but I need to understand better your use-case, can you describe in more detail the end result you expect ?

Reply:
Very long thread :)

IMHO I would split Depict from the Samsung Frame TV.

For Depict, you really just need a camera, you pass the camera URL to Depict and that's it. Don't call it camera.depict, call it camera.webdav. It would be nice to have such camera showing up family pictures in the middle of the interface :)

For samsung frame tv... definitely a media_player, ultimately that's what you're doing...

The issue now is the playlist.... for the camera.webdav, the playlist can be part of the component (that is something IMHO fits into the scope of the "camera"... but don't take my word for it). Another option would be to create a new component just for the purpose of the playlist's...

I think we should take baby steps and learn in the process, why don't you start with the camera.webdav ? I would use it :)